File: 2015/src/days/day1.py
from .day import DayAbstract


class Day(DayAbstract):
    def __init__(self, data: str) -> None:
        super().__init__(data)

    def parse(self, data) -> str:
        return data

    def part_one(self) -> int:
        floor = 0
        
        for char in self.data:
            floor += 1 if char == '(' else -1
        
        return floor

    def part_two(self) -> int:
        floor = 0

        for index, char in enumerate(self.data):
            floor += 1 if char == '(' else -1

            if floor < 0:
                return index + 1
        
        return 0

File: 2022/src/days/day1.ts
import { DayAbstract } from '../utils/day';

type Input = number[][];

export class Day extends DayAbstract {
    parse(input: string): Input {
        return input.split('\n\n').map((line) => line.split('\n').map(Number));
    }

    partOne() {
        return this.input
            .map((numbers: number[]) =>
                numbers.reduce((acc: number, num: number) => acc + num, 0)
            )
            .reduce((max: number, num: number) => (max > num ? max : num), 0);
    }

    partTwo() {
        return this.input
            .map((numbers: number[]) =>
                numbers.reduce((acc: number, num: number) => acc + num, 0)
            )
            .sort((a: number, b: number) => b - a)
            .slice(0, 3)
            .reduce((acc: number, num: number) => acc + num, 0);
    }
}
